demo/app.py

```diff
@@ -35,8 +67,52 @@ def index():
         response['code'] = res.status_code
         response['data'] = json.loads(res.text)['data']
 
-    return render_template('index.html', form=onid_form, onid=onid, res=response)
+    return render_template('index.html', form=onid_form, onid=onid, res=response, logout_url=logout_url)
+
+@app.route('/logout', methods=['GET'])
+def logout():
+    if 'user' in session:
+        del session['user']
+    return flask.redirect(app.config['CAS_URL'] + "/logout")
+
+class CASError(Exception):
+    def __init__(self, msg, code=''):
+        self.msg = msg
+        self.code = code
+    def __str__(self):
+        if self.code and self.msg:
+            return str(self.msg) + " (" + str(self.code) + ")"
+        if self.msg:
+            return str(self.msg)
+        return str(self.code)
+
+def validate_cas(cas_url, ticket, service):
+    r = requests.get(cas_url+'/serviceValidate', params={'ticket': ticket, 'service': service})
+    if r.status_code != 200:
+        raise CASError('invalid response')
+
+    # note: processing xml from untrusted sources is unsafe
+    # https://docs.python.org/2/library/xml.html#xml-vulnerabilities
+    try:
+        root = elementtree.fromstring(r.text)
+    except elementtree.ParseError:
+        raise CASError('invalid response')
+
+    ns = '{http://www.yale.edu/tp/cas}'
```

Comment:
What does this stand for?

Reply:
namespace. I will update it to make that more clear